deeplinkdispatch-base/src/main/java/com/airbnb/deeplinkdispatch/base/MatchIndex.java

```diff
@@ -128,83 +108,109 @@ public Match matchUri(@NonNull List<UrlElement> elements, @Nullable Map<String,
     int valueLength = getValueLength(elementStartPos);
-    if (valueLength > 0 && byteArray[valuePos] == IDX_PLACEHOLDER) {
-      valuePos += 1; // The actual value starts at +1 if we have a placeholder indicator
-      valueLength -= 1; // IF placeholder the value is one byte shorter
-      containsPlaceholder = true;
-      if (valueLength == 2) {
-        return null; // Empty placeholder id does not match
-      }
-      if (value.length == 0) {
-        return null; // Empty string does not match the placeholder
-      }
-    }
-    if (byteArray[elementStartPos] != type || (valueLength != value.length
-        && !containsPlaceholder)) {
-      return null; // Does not match
+    boolean isComponentTypeMismatch = nodeMetadata != inboundUriComponentType;
+    boolean isValueLengthMismatch = valueLength != inboundValue.length;
+
+    if ((isComponentTypeMismatch || isValueLengthMismatch) && isValueLiteralValue) {
+      //Opportunistically skip comparing this node before walking the two values if we can infer
+      // that they will not match.
+      return null;
     }
-    // i index over value array forward
-    // j index over search byte arrays value element forward
-    // k index over value array backward
-    for (int i = 0; i < value.length; i++) {
-      if (containsPlaceholder && byteArray[valuePos + i] == '{') {
-        // Until here every char in front for the placeholder matched.
-        // Now lets see of all chars behind thje placeholer also match
-        for (int j = valueLength - 1, k = value.length - 1; j >= 0; j--, k--) {
-          if (byteArray[valuePos + j] == '}') {
-            // Text behind the placholder fully matches. Now we just need to get the placeholer
-            // string and can return.
-            byte[] placeholderValue = new byte[k - i + 1];
-            // Size is without braces
-            byte[] placeholder = new byte[(valuePos + j) - (valuePos + i) - 1];
-            System.arraycopy(value, i, placeholderValue, 0, placeholderValue.length);
-            System.arraycopy(byteArray, valuePos + i + 1, placeholder, 0, placeholder.length);
-            return new StringBuilder(placeholderValue.length + placeholder.length + 1)
+    if (isComponentParam) {
+      if ((
+        //Per com.airbnb.deeplinkdispatch.DeepLinkEntryTest.testEmptyParametersNameDontMatch
+        //We should not return a match if the param (aka placeholder) is empty.
+        byteArray[valueStartPos] == '{' && byteArray[valueStartPos + 1] == '}'
+      ) || (
+        //Per com.airbnb.deeplinkdispatch.DeepLinkEntryTest.testEmptyPathPresentParams
+        //We expect an empty path to not be a match
+        inboundValue.length == 0
+      )) {
+        return null;
+      }
+      // i index over inboundValue array forward
+      // j index over search byte arrays inboundValue element forward
+      // k index over inboundValue array backward
+      for (int i = 0; i < inboundValue.length; i++) {
```

Comment:
nit: This function is getting long, might be nice to break into two smaller functions, one for  comparing componentParam, and one for configurablePathSegment